src/lib/optimizer/strategy/dependent_group_by_reduction_rule.cpp

```diff
@@ -0,0 +1,132 @@
+#include "dependent_group_by_reduction_rule.hpp"
+
+#include <unordered_map>
+
+#include "expression/abstract_expression.hpp"
+#include "expression/expression_functional.hpp"
+#include "expression/expression_utils.hpp"
+#include "hyrise.hpp"
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "logical_query_plan/aggregate_node.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/projection_node.hpp"
+#include "logical_query_plan/stored_table_node.hpp"
+
+using namespace opossum::expression_functional;  // NOLINT
+
+namespace opossum {
+
+void DependentGroupByReductionRule::apply_to(const std::shared_ptr<AbstractLQPNode>& lqp) const {
+  // Store a copy of the root's column expressions.
+  const auto root_column_expressions = lqp->column_expressions();
+  visit_lqp(lqp, [&](const auto& node) {
+    if (node->type != LQPNodeType::Aggregate) {
+      return LQPVisitation::VisitInputs;
+    }
+    auto& aggregate_node = static_cast<AggregateNode&>(*node);
+
+    std::unordered_map<std::shared_ptr<const StoredTableNode>, std::set<ColumnID>> group_by_columns_per_table;
+    // Collect the group-by columns for each table in the aggregate node
+    for (auto expression_idx = size_t{0}; expression_idx < aggregate_node.aggregate_expressions_begin_idx;
+         ++expression_idx) {
+      const auto& expression = aggregate_node.node_expressions[expression_idx];
+      const auto& column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(expression);
+      if (!column_expression) continue;
+
+      const auto& stored_table_node =
+          std::dynamic_pointer_cast<const StoredTableNode>(column_expression->column_reference.original_node());
+      // If column is not a physical column skip
+      if (!stored_table_node) continue;
```

Comment:
StaticTableNodes also have physical columns

Reply:
Argh ... never heard of them. Is there a standard way to handle this? Shall I check for Static and Stored and then store a `BaseNonQueryNode` ptr?